Userland/Libraries/LibWeb/HTML/Worker.h

```diff
@@ -0,0 +1,168 @@
+    Bindings::WebEngineCustomData m_custom_data;
+
+    NonnullRefPtr<JS::VM> m_worker_vm;
+    NonnullOwnPtr<JS::Interpreter> m_interpreter;
+    OwnPtr<WorkerEnvironmentSettingsObject> m_inner_settings;
+    JS::VM::InterpreterExecutionScope m_interpreter_scope;
+    JS::ExecutionContext m_execution_context;
+    WeakPtr<JS::Realm> m_worker_realm;
+    JS::GlobalObject* m_worker_scope;
+
+    WeakPtr<MessagePort> m_implicit_port;
+    RefPtr<MessagePort> m_outside_port;
+
+    void run_a_worker(AK::URL& url, EnvironmentSettingsObject& outside_settings, MessagePort& outside_port, WorkerOptions const& options);
+};
+
+// FIXME: This is a bit ugly, ESO should simply take the document instead of window.
+class WorkerEnvironmentSettingsObject final : public EnvironmentSettingsObject {
+public:
+    WorkerEnvironmentSettingsObject(DOM::Document& document, JS::ExecutionContext& execution_context)
+        : EnvironmentSettingsObject(execution_context)
+        , m_document(document)
+    {
+    }
+    static NonnullOwnPtr<WorkerEnvironmentSettingsObject> setup(DOM::Document& document, JS::ExecutionContext& execution_context/* FIXME: null or an environment reservedEnvironment, a URL topLevelCreationURL, and an origin topLevelOrigin */)
+    {
+        // 1. Let realm be the value of execution context's Realm component.
+        auto* realm = execution_context.realm;
+        VERIFY(realm);
+
+        // 2. Let window be realm's global object.
+        // NOTE: We want to store the Window impl rather than the WindowObject.
+        //auto& window = verify_cast<Bindings::WindowObject>(realm->global_object()).impl();
+
+        // 3. Let settings object be a new environment settings object whose algorithms are defined as follows:
+        // NOTE: See the functions defined for this class.
+        auto settings_object = adopt_own(*new WorkerEnvironmentSettingsObject(document, execution_context));
+
+        // FIXME: 4. If reservedEnvironment is non-null, then:
+        // FIXME:    1. Set settings object's id to reservedEnvironment's id, target browsing context to reservedEnvironment's target browsing context, and active service worker to reservedEnvironment's active service worker.
+        // FIXME:    2. Set reservedEnvironment's id to the empty string.
+
+        // FIXME: 5. Otherwise, set settings object's id to a new unique opaque string, settings object's target browsing context to null, and settings object's active service worker to null.
+        settings_object->target_browsing_context = nullptr;
+
+        // FIXME: 6. Set settings object's creation URL to creationURL, settings object's top-level creation URL to topLevelCreationURL, and settings object's top-level origin to topLevelOrigin.
+        //settings_object->creation_url = creation_url;
+
+        // 7. Set realm's [[HostDefined]] field to settings object.
+        realm->set_host_defined(move(settings_object));
+
+        return settings_object;
```

Comment:
Here you `move` the settings to the realm which means our local variable `settings_object` is no longer valid.

To fix this:
Make this function return `WeakPtr<WorkerEnvironmentSettingsObject>`
The return below has to be: `static_cast<WorkerEnvironmentSettingsObject*>(realm->host_defined())`
(Maybe with a comment that we just set this so this should be valid at this point)

And in `class Worker` make `m_inner_settings` a `WeakPtr<WorkerEnvironmentSettingsObject>` as well.

To be allowed to make `WeakPtr`s add `, public Weakable<WorkerEnvironmentSettingsObject>` to the extends of `WorkerEnvironmentSettingsObject`